CSC Halo Filter update: HLT and offline

lathomas opened the pull request:
Update of the beam halo filter. 

A. Methods using halo triggers (L1 or HLT) information are not used anymore. 

B. The filter is now based on the presence of a calo deposit compatible with a halo hypothesis. 
B.1 First, halo cluster candidates are built using e.g CSCHaloAlgo::GetHaloClusterCandidateEB for EB clusters.
(e.g. RecoMET/METAlgorithms/interface/HaloClusterCandidateEB.h). 
Some local isolation criteria are required. A loose H/E or E/H cut is also applied in most of the cases. 
The criteria were tuned to reach a cluster efficiency of more than 99% for interacting beam halos. 

B.2 To be flagged as a beam halo deposit, the cluster candidates must either: 
a) Have a pattern compatible with a halo hypothesis (timing conditions, large size in eta in the barrel,...). This pattern finding is done e.g. by the method:
CSCHaloAlgo::EBClusterShapeandTimeStudy.
or 
b) be matched with a CSC segment (dphi, dR, dt conditions). The matching is performed by the method:  CSCHaloAlgo::ApplyMatchingCuts. 

C. Two new versions of the filter have been added to the BeamHaloSummary (in DataFormats/METReco): 
-CSCTightHalo2016: If any cluster is flagged as a halo deposit by either a) OR b), in any region, the event is declared as halo. 
=> The fake rate was estimated on data to be 0.4% (JetHT)/0.2% (MuonEG), for an efficiency of around 96% for true beam halos. 
=> The fake rate and efficiency of the 2015 version of the filter were estimated to be 0.8%/85%. 

-CSCSuperTightHaloId2016:  If any cluster is flagged as a halo deposit by a) AND b), in any region, the event is declared as halo
=> The fake rate is less than 0.01% for an efficiency of 85%. 

D. Variables based on CSC information only such as the size of the longest chain of flat segments and the presence of an opposite endcap pair of flat segments are still stored in CSCHaloData. 
These variables are not used in the standard filter, since passive (i.e. non interacting) halo tracks are believed not to affect the event reconstruction and that the probability to  observe such a track on top of a good collision event is not negligible and could go up to several percents in 2016. 
They could however still be of interest for specific analyses such as heavy stable particles searches, where non interacting halos are an important background. 

E. Rechit cleaning: 
The CSCHaloData also stores a RefVector with all the Ecal and HBHE Rechits associated to a halo cluster.
A new producer CaloRecHitsBeamHaloCleaned (RecoMET/METProducers) was added, that simply produces a list of rechits after beam halo cleaning. 
For now, this cleaning is not performed for the offline reconstruction but will be used by some MET trigers. 

F. HLT version of the filter. 
It was noticed in 2015 that a sizeable fraction of the rate of HLT_PFMET170 was due to beam halos. The situation is expected to get worse this year. 
It is therefore planned to use a cleaning module to get rid of beam halos events. 
The same code as for the offline filter is used but with higher et thresholds in some cases. Conditions based on the rechit time are not considered, as the rechit time at HLT doesn't appear to be reliable.
The performances of the cleaning were measured on three datasets: 
-MET, where half of the events passing the HLT_PFMET170_HBHECleaned trigger are not passing the trigger after cleaning 
-JetHT, where a halo-like cluster is found in only 0.3% of the events, out of which 2/3 are real beam halos. 
-MuonEG, where a halo-like cluster is found in less than 0.1% of the cases. 
The proper trigger sequence will be communicated soon to the TSG group. 

References: 
Presentation at the TSG meeting (March 16th): https://indico.cern.ch/event/504435/contribution/2/attachments/1244481/1832023/beamhalo.pdf
Presentation at the CSC meeting (Jan 13th): https://indico.cern.ch/event/482724/contribution/29/attachments/1211233/1766683/bhreport.pdf

makortel committed:
First version of the new quadruplet algorithm

makortel committed:
Add new algorithm for pixel track cleaner for quadruplets to reduce duplicate rate

makortel committed:
Add customisation to enable new quadruplets

thomreis committed:
Use quality to decide coordinate based cancel out and min pT as a second step when the qualities are equal

thomreis committed:
Prevent warning messages from being printed twice.

Commit:
Merged l1t-calostage2-CMSSW_8_0_0_pre6 from repository cms-l1t-offline

akalinow committed:
* add configuration reset before loading a new configuration.
  Affects reloading configuration as the change of the run.

akalinow committed:
* add configuration reset before loading a new configuration.
  Affects reloading configuration as the change of the run.

akalinow committed:
Merge pull request #154 from akalinow/configButfix

Config butfix

thomreis committed:
Calculate and add input muon index to uGMT output muons

Commit:
Fix jet PUS bugs which would ignore -ve eta, screw up HF jets.

This fixes changes made to add HF into jet finder, in the following
commits:

- 7a72ec8139f0c4d109524377c1f5df0ed5012df8 (pre5 branch)

- 1994746683698e47bf415658294bdd7b1e1bb0cd (pre6 branch)

thomreis committed:
Merged l1t-muon-CMSSW_8_0_0_pre6 from repository cms-l1t-offline

Commit:
Update L1JEC to take into account PUS bugfix.

Also clean up some dud comments in the caloParams configs.

mulhearn committed:
add back support for L1 tech triggers before stage-2

gflouris committed:
Merged l1t-muon-CMSSW_8_0_0_pre6 from repository cms-l1t-offline

gflouris committed:
Extra LUTs for algo2

gflouris committed:
Algo2 phib

gflouris committed:
set bx range

gflouris committed:
WS - three output muons

gflouris committed:
use rpc if dt low quality

gflouris committed:
Merge pull request #158 from gflouris/mydevel_800pre6

BMTF New Algorithm

jimbrooke committed:
unpacker update for demux FED move

jimbrooke committed:
Merge branch 'l1t-calostage2-CMSSW_8_0_0_pre6' of github.com:cms-l1t-offline/cmssw into mydevel

mulhearn committed:
Make Calo Jet/EGamma/Tau packing consistent with unpacking, wrt HW Eta

thomreis committed:
Add forgotten increment for muon index counter

thomreis committed:
Use only 2 MSBs for uGMT output muon quality

thomreis committed:
use only quality based cancel out

mulhearn committed:
force checkout of L1Trigger/L1TGlobal, as a work-around (this is fixed properly in 8_0_0)

Commit:
Fix bug L1RecoTreeProducer where variables not written to tree.

thomreis committed:
Adjust also intermediate muon quality

Commit:
Reset vectors in L1AnalysisRecoVertexDataFormat before each event.

Failure to do this results in massive vectors and filesize.

konec committed:
Merged l1t-muon-CMSSW_8_0_0_pre6 from repository cms-l1t-offline

konec committed:
changes from Pawel towards separation of object, definitely not final

SridharaDasu committed:
Added HF LUTs with unity transform as default

SridharaDasu committed:
Added HFLUTs table

SridharaDasu committed:
Updated HFLUT values - no longer unity

mulhearn committed:
Merged l1t-calostage2-CMSSW_8_0_0_pre6 from repository cms-l1t-offline

Conflicts:
	L1Trigger/L1TCalorimeter/src/firmware/Stage2Layer2JetSumAlgorithmFirmwareImp1.cc

mulhearn committed:
fix crash in runL1T unit test due to missing ESProducer for CaloTPGRecord

mulhearn committed:
Merged l1t-muon-CMSSW_8_0_0_pre6 from repository cms-l1t-offline

konec committed:
changes from Pawel to accomodate recent changes from Artur

thomreis committed:
Merge branch 'l1t-muon-CMSSW_8_0_0_pre6_uGMT-dev-cancelByQuality' into l1t-muon-CMSSW_8_0_0_pre6_uGMT-dev-index-bits

thomreis committed:
Merge pull request #167 from thomreis/l1t-muon-CMSSW_8_0_0_pre6_uGMT-dev

uGMT developments

thomreis committed:
Set good muon quality of QualityAdjuster to 12 so that the two MSBs for the uGT are 1

thomreis committed:
Add TF muon index variable to upgrade muon ntuple

mulhearn committed:
add customizations for Event Setup of HF 1x1 TPs

mulhearn committed:
deprecate reEmul.py for customizations, after bitwise equivalence demonstrated

mulhearn committed:
switch back to stand-in Layer1 by default

SridharaDasu committed:
Merged l1t-integration-v4.0 from repository cms-l1t-offline

mulhearn committed:
deprecate customise_Stage2Calo.py

mulhearn committed:
Merged l1t-muon-CMSSW_8_0_0_pre6 from repository cms-l1t-offline

SridharaDasu committed:
Changed to caloPhi=1-72 for HF as well

jimbrooke committed:
revert HT to sum of all jets no jet ID

jimbrooke committed:
increase default jet threshold for HT, MHT to 35 GeV

gflouris committed:
Eta map fix

mulhearn committed:
Merge branch 'mydevel' into bugfix

mulhearn committed:
Merge branch 'bugfix' into mydevel

mulhearn committed:
revert to stand-in layer-1 emulator

mulhearn committed:
switch to FW version 2.0 for BMTF

mulhearn committed:
L1T unit testing

SridharaDasu committed:
Added region unpacking and fixed test program

SridharaDasu committed:
Now makes region summary from raw data

SridharaDasu committed:
Merged l1t-integration-CMSSW_8_0_0_pre6 from repository cms-l1t-offline

SridharaDasu committed:
Put in HFTPGs in HCALDigis again as that change went away mysteriously

blwiner committed:
Fix for GT muon unpacking

jimbrooke committed:
Merged dasu-updates-for-layer1 from repository cms-l1t-offline

SridharaDasu committed:
Fixed HF 1x1 tower handling

jimbrooke committed:
Merge branch 'dasu-updates-for-layer1' of github.com:cms-l1t-offline/cmssw into from-CMSSW_8_0_0_pre6

mulhearn committed:
Merge branch 'integration' into dasu-updates-for-layer1

mulhearn committed:
Merge branch 'dasu-updates-for-layer1' into integration

mulhearn committed:
fix crash in GtBoard when exit before 1 event processed

mulhearn committed:
fix integration test code

mulhearn committed:
extend Calo objects to include intermediate values from emulation

jimbrooke committed:
set towers in layer 1 output vector, instead of push back

konec committed:
fix quality, now based on hit pattern

thomreis committed:
Merged l1t-tsg-v3 from repository cms-l1t-offline

akalinow committed:
Merge pull request #181 from konec/omtfdev_mk3

New quality definition, and code refactorin

jimbrooke committed:
actually apply the change

gflouris committed:
Merged l1t-muon-quality-CMSSW_8_0_0_pre6 from repository cms-l1t-offline

mulhearn committed:
improvements to integration testing to check L1T global output, unpacking

mulhearn committed:
Merge branch 'l1t-calospeedup-CMSSW_8_0_0_pre6' of https://github.com/jimbrooke/cmssw into jimbrooke-l1t-calospeedup-CMSSW_8_0_0_pre6

mulhearn committed:
Merge branch 'jimbrooke-l1t-calospeedup-CMSSW_8_0_0_pre6' into integration

gflouris committed:
4bit Quality v2

gflouris committed:
Masking

mulhearn committed:
bugfix integration test

mulhearn committed:
switch on Calo Layer1 bitwise emulate by default

mulhearn committed:
Revert "switch on Calo Layer1 bitwise emulate by default"

This reverts commit 5b467c198e9070b8deb67f04285f16e4a8101d26.

mulhearn committed:
add support for convenient command-line driven config files, while starting from cmsDriver.py

mrcarver committed:
Merged l1t-muon-quality-CMSSW_8_0_0_pre6 from repository cms-l1t-offline

mrcarver committed:
Added Correct BX and quality assignment

mrcarver committed:
Added Correct BX and quality assignment

thomreis committed:
uGMT ntuple and config file

thomreis committed:
use TF emulator muons as input for the uGMT istead of the quality adjusted ones

thomreis committed:
output edm root file

mrcarver committed:
Fixed lack of EMTF output muons

mrcarver committed:
Fixed lack of EMTF output muons

mrcarver committed:
Merge branch 'l1t-muon-quality-CMSSW_8_0_0_pre6' of github.com:cms-l1t-offline/cmssw into QualBXFixForThomas

Needed to pull in extra changes of l1t-muon-quality-CMSSW_8_0_0_pre6 branch so I can update branch with EMTF output fix

mulhearn committed:
update to L1T integration tests adding bitwise comparison of Calo/Muon outputs...\n

rekovic committed:
Merged l1t-integration-v9.3 from repository cms-l1t-offline

mulhearn committed:
add menu summary comparison to integration test, until full GT is added to L1TNtuple

mulhearn committed:
Merged l1t-tsg-v3 from repository cms-l1t-offline

mulhearn committed:
backport integration tests to l1t-tsg-v3

mulhearn committed:
Merged cp-backport-integration-tests from repository cms-l1t-offline

Conflicts:
	EventFilter/L1TRawToDigi/src/implementations_stage2/RegionalMuonGMTPacker.cc
	L1Trigger/Configuration/python/SimL1Emulator_cff.py
	L1Trigger/Configuration/python/customiseReEmul.py
	L1Trigger/Configuration/python/customiseUtils.py
	L1Trigger/L1TCommon/python/l1tSummaryStage1Digis_cfi.py
	L1Trigger/L1TCommon/python/l1tSummaryStage2Digis_cfi.py
	L1Trigger/L1TCommon/test/reEmul.py
	L1Trigger/L1TCommon/test/runL1T.py
	L1Trigger/L1TGlobal/src/GtBoard.cc
	L1Trigger/L1TNtuples/python/l1UpgradeTree_cfi.py

mrcarver committed:
Updated BX assignment to reflect using second earliest LCT

mrcarver committed:
Updated BX assignment to reflect using second earliest LCT

rekovic committed:
Merged l1t-integration-v9.4 from repository cms-l1t-offline

davignon committed:
Updated stage2 tau compression, calibration and isolation LUTs, updated tau emulator for new LUTs

SridharaDasu committed:
Fixed LUT usage?

thomreis committed:
fix for ugmt ntuple. correct tower sum input for ugmt

mrcarver committed:
Merge branch 'l1t-muon-quality-CMSSW_8_0_0_pre6' of github.com:cms-l1t-offline/cmssw into QualBXFixForThomas

Pulled in other developments before I fix BX and quality again

mrcarver committed:
Fixed quality assignment that got messed up somehow

mrcarver committed:
Removed inzone requirement from phi matching function to reflect FW

davignon committed:
Updated stage2 tau compression, calibration and isolation LUTs, updated tau emulator for new LUTs - after validation

mulhearn committed:
add comparisons for sim sequence

mulhearn committed:
turn off test failing due to missing file

mulhearn committed:
turn on layer-1

mulhearn committed:
Revert "turn on layer-1"

This reverts commit c93f8db6d2fefad8abfeeb2fcc0f66c2f0875e69.

rekovic committed:
Merge remote-tracking branch 'remotes/cms-l1t-offline/l1t-integration-CMSSW_8_0_0_pre6' into l1t-muon-quality-CMSSW_8_0_0_pre6

rekovic committed:
Revert "fix for ugmt ntuple. correct tower sum input for ugmt"

This reverts commit fdaf95b3afb87e5e1cb5485020e5c0450218cc63.

rekovic committed:
Revert "output edm root file"

This reverts commit aaec59024de20197872f0ccdf95342c9c19377e8.

rekovic committed:
Revert "uGMT ntuple and config file"

This reverts commit 0923aaf720d04688452984cb4188bce2237f0c0d.

rekovic committed:
Use the Trigger Tower Sums

rekovic committed:
Merge branch 'l1t-muon-quality-CMSSW_8_0_0_pre6' into l1t-integration-CMSSW_8_0_0_pre6

VinInn committed:
add index for gdets

mulhearn committed:
udpates tests to add discrepant plots and distinguish forward-jet-only changes

mulhearn committed:
Merged l1t-dev-l1t-tsg-v3-801 from repository cms-l1t-offline

mulhearn committed:
Merge branch 'integration' into l1t-integration-CMSSW_8_0_0_pre6

Conflicts:
	DataFormats/L1Trigger/src/classes_def.xml
	L1Trigger/L1TCommon/macros/NtupleDiff.C
	L1Trigger/L1TCommon/scripts/testL1T.pl
	L1Trigger/L1TGlobal/src/GtBoard.cc

gomber committed:
Changed hcal scale

mulhearn committed:
Merge branch 'l1t-integration-CMSSW_8_0_0_pre6' into integration

mulhearn committed:
bring back diff plots

mulhearn committed:
turn on layer1 by default

mulhearn committed:
turn on layer1 by default

mulhearn committed:
Revert "turn on layer1 by default"

This reverts commit 5eee1cd0282fb91d1b391d5480f8195f9802283c.

mulhearn committed:
Revert "turn on layer1 by default"

This reverts commit bab44b7e22d907c339fdaaba18ec0cb6c27f8e8c.

VinInn committed:
faster, not as much as expected

VinInn committed:
infinitely faster hash

VinInn committed:
cure expensive math

VinInn committed:
ofast

VinInn committed:
exchange tests: slower second

VinInn committed:
speed up EtaPhiMeasurementEstimator

VinInn committed:
modernize EtaPhiRange, remove obsolere GlobalDetRange

makortel committed:
Fixes for clang

VinInn committed:
devirtualize, speed up math

VinInn committed:
make atan2 3bits more precise

mulhearn committed:
Merged l1t-tsg-v4 from repository cms-l1t-offline

mulhearn committed:
purge data area of files put into cms-data

mulhearn committed:
remove PDF files form L1TRawToDigi

mulhearn committed:
cleanup unneeded dummy change to L1TGlobal

mulhearn committed:
remove root file

mulhearn committed:
Merged refs/pull/13767/head from repository cms-sw

Conflicts:
	DataFormats/L1TMuon/interface/EMTFOutput.h
	L1Trigger/L1TMuonOverlap/src/OMTFSorter.cc

VinInn committed:
fix trivial bug, no regression

makortel committed:
Take SeedingLayerSetsHits::SeedingLayerSet via const reference

The object has const interface, and consists of a pointer and two
iterators, so the technical gains are tiny, but it makes the interface
of triplet and quadruplet generators more uniform.

VinInn committed:
modernize and add Ofast where relevant

VinInn committed:
leave both versions, use old one

Sam-Harper committed:
adding first draft of EgammaHLTPixelMatchS2Producer

VinInn committed:
use std::shared_ptr

Sam-Harper committed:
updating to use a TF1 instead of hardcoded functions

Sam-Harper committed:
fill descript should now work

VinInn committed:
change from map to heap

Commit:
Update a test cfi file for testing finer segmentation of BH

Sam-Harper committed:
now fully templated

VinInn committed:
use priority_queue

VinInn committed:
remove debug

VinInn committed:
make DynArray std compatible

nsmith- committed:
Merged pr-l1t-tsg-v4-81x from repository cms-l1t-offline

VinInn committed:
test swa[

VinInn committed:
remove debug

nsmith- committed:
L1TCaloLayer1 move to MessageLogger

nsmith- committed:
Move L1TCaloLayer1 constants to anonymous namespace + constexpr

nsmith- committed:
Add reserves for some std::vectors, std::abs

nsmith- committed:
Move statics to class member variable

VinInn committed:
dummy out hit etaphi estimator

wmtan committed:
Use unique_ptr, not auto_ptr

Commit:
Remove obsolete parameter definitions

cmsbuild committed:
Merge pull request #13809 from cms-l1t-offline/l1t-fixCorrlCond-EtSumVec-80x

Fix crash in CorrelationConditions uGT emulator

Sam-Harper committed:
adding other paramertisation objects

Sam-Harper committed:
a flaw...

VinInn committed:
return components by const reference

Commit:
remove unused parameters

Commit:
More realistic ME0 geometry and extended GE21 geometry

VinInn committed:
add wrapper in other places

VinInn committed:
remove debug

Sam-Harper committed:
working S2 pixel match module

Sam-Harper committed:
fixing product writing control

Sam-Harper committed:
renaming pixel match producer

Commit:
Get geometry information of GEM detector

lathomas committed:
CSC Halo Filter update: HLT and offline

cmsbuild commented:
A new Pull Request was created by  @lathomas  for CMSSW_8_1_X.

It involves the following packages:

DataFormats/METReco
PhysicsTools/PatAlgos
RecoMET/METAlgorithms
RecoMET/METFilters
RecoMET/METProducers


@cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please review it and eventually sign? Thanks.
@TaiSakuma, @imarches, @ahinzmann, @acaudron, @mmarionncern, @rappoccio, @jdolen, @nhanvtran, @schoef, @ferencek, @gpetruc, @mariadalfonso, @pvmulder this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

cmsbuild committed:
Merge pull request #13799 from bsunanda/Phase2-hgx46

bsunanda:Phase2-hgx46 Update a test cfi file for testing finer segmentation of BH

davidlange6 committed:
Merge pull request #13820 from wmtan/UseUniquePtr

Use unique_ptr, not auto_ptr

slava77 commented:
@lathomas 
were the presentations in TSG and CSC meetings corresponding to the same code as is submitted here?
Jan 13 seems pretty far in the past.

slava77 commented on DataFormats/METReco/interface/CSCHaloData.h:
```diff
@@ -69,6 +78,20 @@ namespace reco {
     edm::RefVector<reco::TrackCollection>& GetTracks(){return TheTrackRefs;}
     const edm::RefVector<reco::TrackCollection>& GetTracks()const {return TheTrackRefs;}
     
+    edm::RefVector<EcalRecHitCollection>& GetEBRechits(){return ecalebrhRefs;}
```
It doesn't seem appropriate to extend CSCHaloData with calorimeter information.
The design of the halo data formats was to collect detector-specific information first and then have a combination in GlobalHaloData (or in BeamHaloSummary)

slava77 commented on DataFormats/METReco/interface/CSCHaloData.h:
```diff
@@ -69,6 +78,20 @@ namespace reco {
+
+    edm::RefVector<HBHERecHitCollection>& GetHERechits(){return herhRefs;}
+    const edm::RefVector<HBHERecHitCollection>& GetHERechits()const {return herhRefs;}
+    */
```
Is this commented out piece needed?
Please remove or add a comment why it is here.

slava77 commented on RecoMET/METAlgorithms/interface/CSCHaloAlgo.h:
```diff
@@ -91,7 +99,7 @@ class CSCHaloAlgo {
  public:
   CSCHaloAlgo();
-  ~CSCHaloAlgo(){}
+  ~CSCHaloAlgo(){};
```
";" is not needed

slava77 commented on RecoMET/METAlgorithms/interface/CSCHaloAlgo.h:
```diff
@@ -148,17 +156,33 @@ class CSCHaloAlgo {
+  float  et_thresh_rh_hb, dphi_thresh_segvsrh_hb,dr_lowthresh_segvsrh_hb, dr_highthresh_segvsrh_hb, dt_lowthresh_segvsrh_hb, dt_highthresh_segvsrh_hb, dt_segvsrh_hb;
+  float  et_thresh_rh_he, dphi_thresh_segvsrh_he,dr_lowthresh_segvsrh_he, dr_highthresh_segvsrh_he, dt_lowthresh_segvsrh_he, dt_highthresh_segvsrh_he, dt_segvsrh_he;
+  float  et_thresh_rh_eb, dphi_thresh_segvsrh_eb,dr_lowthresh_segvsrh_eb, dr_highthresh_segvsrh_eb, dt_lowthresh_segvsrh_eb, dt_highthresh_segvsrh_eb, dt_segvsrh_eb;
+  float  et_thresh_rh_ee, dphi_thresh_segvsrh_ee,dr_lowthresh_segvsrh_ee, dr_highthresh_segvsrh_ee, dt_lowthresh_segvsrh_ee, dt_highthresh_segvsrh_ee, dt_segvsrh_ee;
```
please add some line-breaks for readability

slava77 commented on RecoMET/METAlgorithms/interface/HaloClusterCandidateEB.h:
```diff
@@ -0,0 +1,97 @@
+
+  
+
+  double GetClusterEt(){return et;}
```
CMSSW style is for class methods to start with lower case letters.
First capital is reserved for type names.
This is a new code. So, it better follow the style from the start.

slava77 replied:
given the implementation, it makes sense to make all these member methods const qualified.
``` double getClusterEt() const ```

slava77 commented on RecoMET/METAlgorithms/interface/HaloClusterCandidateEB.h:
```diff
@@ -0,0 +1,97 @@
+    seed_phi =0;
+    seed_Z =0;
+    seed_R =0;
+    seed_time =0;
```
the amount of copy-paste replicated code would reduce substantially if seed {et,eta, phi, z, r, time} are wrapped in a separate class.

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -35,29 +39,43 @@ CSCHaloAlgo::CSCHaloAlgo()
+  dphi_thresh_segvsrh_hbhe=0.05;
+  dr_lowthresh_segvsrh_hbhe=-100;
+  dr_highthresh_segvsrh_hbhe=20;
+  dt_highthresh_segvsrh_hbhe=30;
```
are these parameters set from some first principles and will never change?
If some need for a change is possible/planned, the parameters should be derived from configuration.

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -662,12 +482,11 @@ reco::CSCHaloData CSCHaloAlgo::Calculate(const CSCGeometry& TheCSCGeometry,
-	      //&& abs(jR - iR) <= max_segment_r_diff 
-	      && (abs(jR - iR)<0.03*abs(jZ - iZ))
+
+	 if ( abs(deltaPhi(jPhi , iPhi)) <=max_segment_phi_diff 
```
std::abs

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -735,17 +553,81 @@ reco::CSCHaloData CSCHaloAlgo::Calculate(const CSCGeometry& TheCSCGeometry,
+   TheCSCHaloData.SetHaloPatternFoundHE(HaloPatternFoundInHE);
+
+   /*
+   //To be removed, just for tests
```
why not remove it now already?

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+    if(haloclustercands[i].GetNbTowersInEta()<3) continue;
+    //Isolation criteria for very short eta strips
+    if(haloclustercands[i].GetNbTowersInEta()==3 && (haloclustercands[i].GetEtStripPhiSeedPlus1()>0.1 || haloclustercands[i].GetEtStripPhiSeedMinus1()>0.1) ) continue;
+    if(haloclustercands[i].GetNbTowersInEta()<=5 && (haloclustercands[i].GetEtStripPhiSeedPlus1()>0.1 && haloclustercands[i].GetEtStripPhiSeedMinus1()>0.1) ) continue;
```
some line-breaks would be useful for readability

slava77 replied:
the c-style for loop  can be changed to range loop ``` for (hcand : haloclustercands ) { ``` 
This will reduce amount of text and potential multiple calls to  ``` haloclustercands[i] ```

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+
+      double etaj = rhposj.eta();
+      double phij = rhposj.phi();
+      if(abs(eta-etaj)>0.2) continue;//This means +/-11 crystals in eta 
```
std::abs here and everywhere

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+    for(size_t jhit = 0; jhit<hbherechitcoll->size(); ++ jhit){
+      const HBHERecHit & rechitj = (*hbherechitcoll)[ jhit ];
+      math::XYZPoint rhposj = getPosition(rechitj.id(),vtx);
+      double rhetj = rechitj.energy()/cosh(rhposj.eta());
```
``` sqrt(rhposj.mag2()/rhposj.perp2()) ``` is a less numerically challenging operation than a chain of exp,log,tan,atan used in cosh(eta())

alja committed:
Move appex of clipping cone in 3D region view.

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+      //This targets IT beam halo (t around - 1ns) 
+      if(rhtj>5){
+      double corrt_j = rhtj + sqrt(rhposj.x()*rhposj.x()+rhposj.y()*rhposj.y() + 320.*320.)/30. - 320./30.;
+      timediscriminator+=log( TMath::Gaus(corrt_j,0,0.4,false)/TMath::Gaus(corrt_j,0.3,0.4,false) );
```
Please reduce to the 2nd order polynomial of corrt_j.
There is no need to do ``` log (exp ( - corrt_j*corrt_j/0.32) ) ``` just to loose precision and possibly generate NaNs on the way.

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+      //Ask the hits to be in the same endcap
+      if(rhposj.z()*rhpos.z()<0)continue;
+      //Selects only second HCAL layer
+      if(abs(rhposj.z())<425) continue;
```
can this be extracted from hit detId instead?
Magic 425 better be derived from something

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+      double phij = rhposj.phi();
+      if(abs(deltaPhi(phi,phij))>0.4 ) continue;
+      
+      double rhrj = sqrt(rhposj.x()*rhposj.x()+rhposj.y()*rhposj.y());
```
is this just  ``` rhposj.perp() ``` ?

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+      double phij = rhposj.phi();
+      if(abs(eta-etaj)>0.4) continue;//This means +/-4 towers in eta 
+      if(abs(deltaPhi(phi,phij))>0.2) continue;//This means +/-2 towers in phi 
+      if(abs(deltaPhi(phi,phij))>0.1&&abs(eta-etaj)<0.2){isiso=false;break;}//The strip should be isolated
```
please don't copy-paste function calls. (The compiler may or may not optimize these away. )
Define a temporary variable instead.

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+      //The best fitted value for R leads to 240 cm (IT) and 330 cm (OT)
+      double rhtj = rechitj.time();
+      timediscriminatorITBH+= TMath::Log10( rhetj )* ( rhtj +0.5*(sqrt(240.*240.+rhposj.z()*rhposj.z()) -abs(rhposj.z()))/30);
+      if(abs(rhposj.z())<300) timediscriminatorOTBH+= TMath::Log10( rhetj )* ( rhtj -0.5*(25-(sqrt(330.*330.+rhposj.z()*rhposj.z()) +abs(rhposj.z()))/30) );
```
please use std::log10

Here and elsewhere in the file, the repeated magic variables should be defined as constants with some intuitive names

slava77 replied:
HCAL reco has a large fraction of ~0 energy hits.
I can't recall if exact zeros still show up.
Please add a check for the hits to be above some small threshold (unless I missed it in upstream selections)

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+      double etaj = rhposj.eta();
+      double phij = rhposj.phi();
+      double dr = sqrt(abs(eta-etaj)*abs(eta-etaj)+deltaPhi(phi,phij)*deltaPhi(phi,phij));
+      if(dr>0.3) continue;
```
there is no need to add sqrt here.
Also abs is not needed in abs*abs expression.

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+}
+
+
+bool CSCHaloAlgo::ApplyMatchingCuts(TString subdet, bool ishlt, double rhet, double segZ, double rhZ, double segR, double rhR, double segT, double rhT, double segPhi, double rhPhi){
```
instead of TString, please define enum.

slava77 replied:
I see that this method is used only with the same inputs from haloclustercands.
It may be more compact to pass a haloclustercand by const reference, instead of extracting the same 5 data members in multiple places.

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+  bool rhmatchingfound =false;
+  for(size_t i =0; i<haloclustercands.size(); i++){
+   
+    if(!ApplyMatchingCuts("EE",ishlt, haloclustercands[i].GetSeedEt(), iZ, haloclustercands[i].GetSeedZ(),iR, haloclustercands[i].GetSeedR(), iT,haloclustercands[i].GetSeedTime(), iPhi, haloclustercands[i].GetSeedPhi()))continue;
```
here and below please add some line breaks for readability

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+
+bool CSCHaloAlgo::ApplyMatchingCuts(TString subdet, bool ishlt, double rhet, double segZ, double rhZ, double segR, double rhR, double segT, double rhT, double segPhi, double rhPhi){
+  //Absolute time wrt BX
+  double tBXrh = rhT+sqrt(rhR*rhR+rhZ*rhZ)/30;
```
define a constant for "30". I suppose it's cmPerNs.

slava77 commented on RecoMET/METFilters/plugins/CSCSuperTightHalo2016Filter.cc:
```diff
@@ -0,0 +1,43 @@
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "DataFormats/METReco/interface/BeamHaloSummary.h"
+
+class CSCSuperTightHalo2016Filter : public edm::global::EDFilter<> {
```
CSCSuperTightHalo2016Filter and CSCTightHalo2016Filter can be merged into more flexible filter module.

slava77 commented on RecoMET/METFilters/python/CSCSuperTightHalo2016Filter_cfi.py:
```diff
@@ -0,0 +1,6 @@
+import FWCore.ParameterSet.Config as cms
+
+CSCSuperTightHalo2016Filter = cms.EDFilter(
```
module instances should start with a lower case letter

slava77 commented on RecoMET/METFilters/python/metFilters_cff.py:
```diff
@@ -73,10 +79,10 @@
 #   hcalLaserEventFilter *
 #  CSCTightHaloTrkMuUnvetoFilter *
-# CSCTightHalo2015Filter *
+   CSCTightHalo2016Filter *
```
is this filter also valid for Run1 and 2015 data (re)processing?
If it is, the 2016 will be confusing.

slava77 commented on RecoMET/METProducers/src/CaloRecHitsBeamHaloCleaned.cc:
```diff
@@ -0,0 +1,349 @@
+
+
+  CSCHaloAlgo CSCAlgo;
+  MuonSegmentMatcher * TheMatcher;
```
CMSSW style is to start data members names with lower case.
Presence of "the" "m_" or trailing underscore is encouraged to differentiate data members from plain local variables in method implementations.

slava77 commented on RecoMET/METProducers/src/CaloRecHitsBeamHaloCleaned.cc:
```diff
@@ -0,0 +1,349 @@
+       (TheSummaryHalo.GetEBRechits()).size()<<", "<<
+       (TheSummaryHalo.GetEERechits()).size()<< ", "<<
+       (TheSummaryHalo.GetHBHERechits()).size()<<", "<<
+       endl;
```
this looks like debug and should either be removed or changed to LogDebug

slava77 commented on RecoMET/METProducers/src/CaloRecHitsBeamHaloCleaned.cc:
```diff
@@ -0,0 +1,349 @@
+void
+CaloRecHitsBeamHaloCleaned::beginStream(edm::StreamID)
+{
+}
```
this and other empty methods can be removed from implementation.

slava77 commented on RecoMET/METProducers/src/CaloRecHitsBeamHaloCleaned.cc:
```diff
@@ -0,0 +1,349 @@
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
+  descriptions.addDefault(desc);
```
it would be much better to actually fill this method with the config parameters and even set their defaults.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: 273503df659e1426f7e9358f8913c416d4c60b4b
 When I ran the RelVals I found an error in the following worklfows: 
 1306.0 step3
<pre>runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15/step3_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>
134.911 step3
<pre>runTheMatrix-results/134.911_RunSinglePh2015D+RunSinglePh2015D+HLTDR2_25ns+RECODR2_25nsreHLT+HARVESTDR2_25nsreHLT/step3_RunSinglePh2015D+RunSinglePh2015D+HLTDR2_25ns+RECODR2_25nsreHLT+HARVESTDR2_25nsreHLT.log

</pre>
1330.0 step3
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15/step3_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>
135.4 step3
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step3_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>
50202.0 step3
<pre>runTheMatrix-results/50202.0_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50/step3_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50.log

</pre>
25202.0 step3
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step3_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

</pre>

 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13849/12081/summary.html 

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/46718602e57f2f6895c4df345cff15bd1fff6094
https://github.com/cms-sw/cmssw/commit/cae5e3b75421d197bf316bb11effffd80e80b6a6
https://github.com/cms-sw/cmssw/commit/0025a309b2fa03985067ab2b6df1a6f6ad8bc30d
https://github.com/cms-sw/cmssw/commit/5e653735059ef68a22a5943fd8ce8c20c69ebb90
https://github.com/cms-sw/cmssw/commit/10d3272210e6e1c5b5aabbab21dc40bfbdef1be7
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13849/12081/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13849/12081/git-merge-result

slava77 commented:
```
1592, in prepare_PATFILTER
    self.schedule.append(getattr(self.process,'Flag_'+filt))
AttributeError: 'Process' object has no attribute 'Flag_CSCTightHalo2016Filter'
```

Please test with ``` runTheMatrix.py -l limited -i all ``` 
This is required for all non-trivial PRs, BTW.
http://cms-sw.github.io/PRWorkflow

davidlange6 committed:
Merge pull request #13850 from alja/clippAppexOffset-8x

8x Fireworks: move clipping cone apex in 3D region view

cmsbuild committed:
Merge pull request #13753 from makortel/new_quadruplets_v4

Quadruplet seeding by propagating triplet to 4th layer

cmsbuild committed:
Merge CMSSW_8_0_X into CMSSW_8_1_X.

nsmith- committed:
Move all L1TCaloLayer1 constants to namespace

Couldn't do `use namespace ...` because constants used in headers

Martin-Grunewald committed:
Less verbose HLTPrescaleProvider

Martin-Grunewald committed:
count[] becomes count_[]

Martin-Grunewald committed:
reorder

Martin-Grunewald committed:
Chris wants unsigned char

wmtan committed:
Change auto_ptr to unique_ptr in text

Commit:
adding dielectron triggers to EXO DQM

Dr15Jones committed:
mv test record.cppunit.cpp from DataFormats/FWLite to PhysicsTools/CondLiteIO

In order to remove dependencies of DataFormats/FWLite on PhysicsTools/CondLiteIO
we needed to move the test of the lite EventSetup system. The test needs
PhysicsTools/CondLiteIO to create the test file.

alja committed:
Set classic root gui style.

davidlange6 committed:
Merge pull request #13859 from Dr15Jones/mvFWLiteEventSetupTestToPhysicsTools

mv test record.cppunit.cpp from DataFormats/FWLite to PhysicsTools/Co…

davidlange6 committed:
Merge pull request #13855 from wmtan/AutoPtrToUniquePtrInText

Change auto_ptr to unique_ptr in text

davidlange6 committed:
Merge pull request #13852 from Martin-Grunewald/HLTPrescaleProviderVerbosity

Less verbose HLTPrescaleProvider (81X)

davidlange6 committed:
Merge pull request #13848 from bsunanda/Phase2-gem12

bsunanda:Phase2-gem12 Get geometry information of GEM detector

davidlange6 committed:
Merge pull request #13860 from alja/guistyleclassic

80x Fireworks: Set classic root gui style

davidlange6 committed:
Merge pull request #13822 from bsunanda/Run2-hcx76

bsunanda:Run2-hcx76 Remove obsolete parameter definitions for relabelling at DIGI step

davidlange6 committed:
Merge pull request #13846 from Sam-Harper/EgPMS2Producer8X

E/g HLT Pixel Match Variable Producer

akhukhun committed:
fix transcoder crash.

smuzaffar committed:
move core_20* generated code and data files in to CSCTrackFinderEmulation external

smuzaffar committed:
L1Trigger/CSCTrackFinder/data files to CSCTrackFinderEmulation external

mulhearn committed:
Merge branch 'nsmith-layer1-style-81X' into pr-l1t-tsg-v4-81x

davidlange6 committed:
add some phase2 (sim only) workflows to the default matrix

mulhearn committed:
remove self reference in library BuildFile

Sam-Harper committed:
fixes for the auto->unique ptr central migration

mulhearn committed:
first step in migration from static variables in OMTF configuration

Commit:
Include HLT_PFMET170_NotCleaned_v

smuzaffar committed:
remove unused testRandomNumberGenerators binary

Commit:
Update the ganging for BH as discussed in the simulation meeting

davidlange6 committed:
Merge pull request #13869 from Sam-Harper/EgPMS2Producer81X

Eg HLT PM module fixes for the auto->unique ptr central migration

Commit:
Resolve conflicts

Commit:
Correct one more xml file

Commit:
Correct one more py file

davidlange6 committed:
Merge pull request #13870 from smuzaffar/testRandomNumberGenerators-cleanup

remove unused testRandomNumberGenerators binary

cmsbuild committed:
Merge pull request #13857 from jruizvar/addExoPaths81X

Adding double-electron triggers to EXO DQM (81X)

Sam-Harper committed:
removing jetcoreregional tracks from tracker isolation sum

mulhearn committed:
complete migration from static variable in OMTF configuration

mulhearn committed:
make static config member variable const in BMTF config

mulhearn committed:
only consume things we are actually going to eat

davidlange6 committed:
Merge pull request #13768 from cms-l1t-offline/pr-l1t-tsg-v4-81x

Bring L1T emulation up to date with l1t-tsg-v4 tag

makortel committed:
Fix compilation under EDM_ML_DEBUG

VinInn committed:
remove debug, add comments

VinInn committed:
remove cout, add comment

wmtan committed:
Use std::shared_ptr and std::make_shared

wmtan committed:
use std::shared_ptr and std::make_shared

davidlange6 committed:
Merge pull request #13885 from makortel/fixDebug

Fix compilation under EDM_ML_DEBUG

davidlange6 committed:
Merge pull request #13873 from Sam-Harper/HEEPTrkIsoFix81X

Bug fix for Electron Tracker Isolation used by HEEP

davidlange6 committed:
Merge pull request #13872 from bsunanda/Phase2-hgx47

bsunanda:Phase2-hgx47 Update the ganging for BH as discussed in the simulation meeting

cmsbuild committed:
Merge pull request #13834 from bsunanda/Phase2-gem11

bsunanda:Phase2-gem11 More realistic ME0 geometry and extended GE21 geometry

davidlange6 committed:
Merge pull request #13867 from davidlange6/phase2sim

add some phase2 (sim only) workflows to the default matrix

davidlange6 committed:
Merge pull request #13868 from akhukhun/fix-hf-lut-crash-81X

fix HF TP transcoder crash.

rekovic committed:
Throw exception if empty or unknown seed in L1TSeedsLogicalExpression

rekovic committed:
Implement L1GlobalDecision functionality: pass without seeds when global OR accept.

slava77 commented:
@lathomas please clarify on your plans to update this PR.
Thank you.

davidlange6 committed:
Merge pull request #13897 from wmtan/StdSharedPtrInGeometry

use std::shared_ptr and std::make_shared (EventSetup in Geometry/Alignment/Visualization)

davidlange6 committed:
Merge pull request #13896 from wmtan/StdSharedPtrInFastSim

Use std::shared_ptr and std::make_shared (EventSetup in FastSim)

davidlange6 committed:
Merge pull request #13866 from smuzaffar/L1Trigger-CSCTrackFinder-to-external

Move L1Trigger/CSCTrackFinder/src/core_ files to external package

davidlange6 committed:
Merge pull request #13831 from VinInn/KullbackLeibler

Speedup Gsf component merging, return components by const reference

davidlange6 committed:
Merge pull request #13819 from VinInn/SpeedHLT81

Speed seeding for HLT

cmsbuild committed:
Merge pull request #13903 from cms-l1t-offline/hlt-l1tseed-update-80x

HLT L1T interface updates - trhow exceptions and L1GlobalDecision  80x

lathomas committed:
Merge branch 'CMSSW_8_1_X' of https://github.com/lathomas/cmssw into CMSSW_8_1_X

cmsbuild commented:
Pull request #13849 was updated. @smuzaffar, @civanch, @Dr15Jones, @lveldere, @rekovic, @cvuosalo, @ianna, @mdhildreth, @fabozzi, @cmsbuild, @alja, @srimanob, @cerminar, @franzoni, @slava77, @ggovi, @mmusich, @hengne, @mulhearn, @ssekmen, @davidlange6 can you please check and sign again.

slava77 commented:
-1

please rebase your PR topic branch to have only relevant commits in the history
(go back to the old release area and do "git rebase XXX" where XXX can be the last IB e.g CMSSW_8_1_X_2016-04-04-1100
